kvbc/src/categorization/immutable_kv_category.cpp

```diff
@@ -120,14 +131,100 @@ void ImmutableKeyValueCategory::deleteBlock(const ImmutableUpdatesInfo &updates_
   }
 }
 
-std::optional<ImmutableValue> ImmutableKeyValueCategory::get(const std::string &key) const {
-  const auto ser = db_->get(cf_, key);
+std::optional<Value> ImmutableKeyValueCategory::get(const std::string &key, BlockId block_id) const {
+  auto val = getLatest(key);
+  if (!val) {
+    return std::nullopt;
+  }
+  if (asImmutable(val).block_id != block_id) {
+    return std::nullopt;
+  }
+  return val;
+}
+
+std::optional<Value> ImmutableKeyValueCategory::getLatest(const std::string &key) const {
+  const auto ser = db_->getSlice(cf_, key);
   if (!ser) {
     return std::nullopt;
   }
-  auto value = ImmutableValue{};
-  deserialize(*ser, value);
-  return value;
+  return value(*ser);
+}
```

Comment:
very nit and personal opinion: I started disliking using `*` for optional may cause confusion when reading and might be more error-prone I think

Reply:
I use it explicitly here, because I first check if there is a value in `ser`. Then, using `*` doesn't do any checks at runtime. In contrast, `value()` does one more check and throws if value is not there. I want to avoid the second check.

Reply:
I did not know that! Thanks for the info @dartdart26  I avoided it in the past for clarity reasons similar to @cloudnoize. But if we can avoid unnecessary checks I think it makes sense to use it.

Reply:
yeah, didn't know that. cool